concourse/model/traits/publish.py

```diff
@@ -137,6 +149,21 @@ def resource_name(self):
     def _children(self):
         return ()
 
+    def validate(self):
+        super().validate()
+        if labels := self.component_descriptor_labels():
```

Comment:
stylistic nit: you might save one level of indentation like so:
```
for label in labels or ():
 ...
```
+ if keeping the approach of checking for a truthy value, it would be better (stricter) to explicitly check for the value being `None` (because, a scalar, false value (such as `false`, or `0`) would be invalid in this ctx).

+ if keeping the if, I would prefer an early-exit (thus also reducing the nesting by one)